internal/store/selector/field.go

```diff
@@ -0,0 +1,384 @@
+package selector
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/flightctl/flightctl/internal/flterrors"
+	"github.com/flightctl/flightctl/pkg/queryparser"
+	"github.com/flightctl/flightctl/pkg/queryparser/sql"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/selection"
+)
+
+type fieldSelector struct {
+	parser        queryparser.Parser
+	fieldResolver *selectorFieldResolver
+}
+
+func NewFieldSelector(dest any) (*fieldSelector, error) {
+	fs := &fieldSelector{}
+
+	var err error
+	fs.fieldResolver, err = SelectorFieldResolver(dest)
+	if err != nil {
+		return nil, err
+	}
+
+	fs.parser, err = sql.NewSQLParser(
+		sql.WithTokenizer(fs),
+		sql.WithOverrideFunction("K", sql.Wrap(fs.queryField)),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return fs, nil
+}
+
+// ParseFromString parses the selector string and returns a SQL query with parameters.
+func (fs *fieldSelector) ParseFromString(ctx context.Context, input string) (string, []any, error) {
+	selector, err := fields.ParseSelector(input)
+	if err != nil {
+		return "", nil, NewSelectorError(flterrors.ErrFieldSelectorSyntax, err)
+	}
+
+	q, args, err := fs.parser.Parse(ctx, selector)
+	if err != nil {
+		if ok := IsSelectorError(err); ok {
+			return "", nil, err
+		}
+		return "", nil, NewSelectorError(flterrors.ErrFieldSelectorParseFailed, err)
+	}
+	return q, args, nil
+}
+
+// Parse parses the selector and returns a SQL query with parameters.
+func (fs *fieldSelector) Parse(ctx context.Context, selector fields.Selector) (string, []any, error) {
+	q, args, err := fs.parser.Parse(ctx, selector)
+	if err != nil {
+		if ok := IsSelectorError(err); ok {
+			return "", nil, err
+		}
+		return "", nil, NewSelectorError(flterrors.ErrFieldSelectorParseFailed, err)
+	}
+	return q, args, nil
+}
+
+// Tokenize converts a selector string into a set of queryparser tokens.
+func (fs *fieldSelector) Tokenize(ctx context.Context, input any) (queryparser.TokenSet, error) {
+	if input == "" {
+		return nil, nil
+	}
+
+	// Assert that input is a selector
+	selector, ok := input.(fields.Selector)
+	if !ok {
+		return nil, fmt.Errorf("invalid input type: expected fieldSelector, got %T", input)
+	}
+
+	requirements := selector.Requirements()
+	tokens := make(queryparser.TokenSet, 0)
+
+	for _, req := range requirements {
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
+
+		field, value, operator := SelectorFieldName(strings.TrimSpace(req.Field)), req.Value, req.Operator
+		resolvedFields, err := fs.fieldResolver.ResolveFields(field)
+		if err != nil {
+			return nil, err
+		}
+
+		resolvedTokens := queryparser.NewTokenSet()
+		for _, resolvedField := range resolvedFields {
+			fieldToken, err := fs.createFieldToken(resolvedField)
+			if err != nil {
+				return nil, err
+			}
+
+			var valueToken *queryparser.TokenSet
+			if value != "" {
```

Comment:
so if value =="" then valueToken is nil? Does createOperatorToken assume it is not?

Reply:
The createOperatorToken requires a tokenset representing the value token.
It doesn't care if the token set is empty or nil, the tokenset is able to process it:
https://github.com/flightctl/flightctl/blob/9e7765f7d44602c5757d1414be45ebc3094a858e/internal/store/selector/field.go#L201

For example, look at the enhance operators branch that implements the entire set of operators.
Since we can have multiple values (i.e., IN), the createOperatorToken still takes a single token set for all values.
https://github.com/flightctl/flightctl/blob/19198b6c7d1a53a2601924e7ce2065f1295cf23b/internal/store/selector/field.go#L111